src/test/java/com/google/cloud/tools/app/impl/cloudsdk/impl/util/FileUtilTest.java

```diff
@@ -0,0 +1,118 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.app.impl.cloudsdk.impl.util;
+
+import static org.junit.Assume.assumeTrue;
+
+import com.google.cloud.tools.app.impl.cloudsdk.internal.util.FileUtil;
+import com.google.common.collect.Sets;
+
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.rules.TemporaryFolder;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.PosixFilePermission;
+import java.util.Set;
+
+/**
+ * Test for {@link FileUtil}
+ */
+public class FileUtilTest {
+
+  @Rule
+  public TemporaryFolder testDir = new TemporaryFolder();
+
+  @Rule
+  public ExpectedException thrown = ExpectedException.none();
+
+  @Test
+  public void testCopyDirectory_nested() throws IOException {
+    Path src = testDir.newFolder("src").toPath();
+    Path dest = testDir.newFolder("dest").toPath();
+
+    Path rootFile = Files.createFile(src.resolve("root.file"));
+    Path subDir = Files.createDirectory(src.resolve("sub"));
+    Path subFile = Files.createFile(subDir.resolve("sub.file"));
+
+    FileUtil.copyDirectory(src, dest);
+
+    Assert.assertTrue(Files.isRegularFile(dest.resolve(src.relativize(rootFile))));
+    Assert.assertTrue(Files.isDirectory(dest.resolve(src.relativize(subDir))));
+    Assert.assertTrue(Files.isRegularFile(dest.resolve(src.relativize(subFile))));
+  }
+
+  @Test
+  public void testCopyDirectory_posixPermissions() throws IOException {
+    assumeTrue(!System.getProperty("os.name").startsWith("Windows"));
+
+    Set<PosixFilePermission> permission = Sets.newHashSet();
+    permission.add(PosixFilePermission.OWNER_READ);
+    permission.add(PosixFilePermission.GROUP_READ);
+    permission.add(PosixFilePermission.OTHERS_READ);
+    permission.add(PosixFilePermission.OTHERS_EXECUTE);
+    permission.add(PosixFilePermission.OTHERS_WRITE);
+
+    Path src = testDir.newFolder("src").toPath();
+    Path dest = testDir.newFolder("dest").toPath();
+
+    Path rootFile = Files.createFile(src.resolve("root1.file"));
+    Assert.assertNotEquals("This test is useless - modified permissions are default permissions",
+        Files.getPosixFilePermissions(rootFile), permission);
+    Files.setPosixFilePermissions(rootFile, permission);
+
+    FileUtil.copyDirectory(src, dest);
+
+    Assert.assertEquals(permission,
+        Files.getPosixFilePermissions(dest.resolve(src.relativize(rootFile))));
+  }
+
+  @Test
+  public void testCopyDirectory_aclPermissions() throws IOException {
+    assumeTrue(System.getProperty("os.name").startsWith("Windows"));
```

Comment:
The problem here lies in the fact that it's a static utility method and I'm testing it as a black box.

I can either take it out or leave it in (and I'm leaning towards leaving it in). Mocking isn't a option here given the style, making it a singleton without state doesn't feel right either.